crates/attestation/src/crypto/crl.rs

```diff
@@ -0,0 +1,341 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+use crate::crypto::SubjectPublicKeyInfoExt;
+
+use anyhow::{bail, Result};
+use der::Encode;
+use x509::crl::CertificateList;
+use x509::PkiPath;
+
+pub trait PkiPathCRLCheck<'a> {
+    fn check_crl(&self, crls: &'a [CertificateList<'a>]) -> Result<()>;
+}
+
+impl<'a> PkiPathCRLCheck<'a> for PkiPath<'a> {
+    fn check_crl(&self, crls: &'a [CertificateList<'a>]) -> Result<()> {
+        // We want to ensure that a valid CRL was passed, so we make sure at least one of the CRLs
+        // is valid for this `PkiPath`, otherwise, no valid CRLs were received, and that's not okay.
+        let mut found = false;
+        let mut iter = self.iter().peekable();
+        loop {
+            let cert = iter.next();
+            let next = iter.peek();
+            if cert.is_none() || next.is_none() {
+                break;
+            }
+            let cert = cert.unwrap();
+            let next = next.unwrap();
+            for crl in crls {
+                if crl.signature_algorithm.oid != cert.signature_algorithm.oid {
+                    // Must not be the correct Certificate-CRL match, go to the next Certificate
+                    // since the signature algorithm on the CRL should match the algorithm
+                    // type of the signing CA, right?
+                    continue;
+                }
+
+                // Don't check CRL expiration in unit tests
+                #[cfg(not(debug_assertions))]
+                if let Some(next_update) = crl.tbs_cert_list.next_update {
+                    if next_update.to_system_time() <= std::time::SystemTime::now() {
+                        bail!("CRL expired");
+                    }
+                }
+
+                let raw_bytes = crl.tbs_cert_list.to_vec().unwrap();
+                match cert.tbs_certificate.subject_public_key_info.verify(
+                    &raw_bytes,
+                    cert.signature_algorithm,
+                    crl.signature.raw_bytes(),
+                ) {
+                    Ok(_) => {
+                        if let Some(revocations) = crl.tbs_cert_list.revoked_certificates.as_ref() {
+                            for revoked in revocations {
+                                if revoked.serial_number == next.tbs_certificate.serial_number {
+                                    bail!("revoked!");
+                                }
+                            }
+                        }
+                        found = true;
```

Comment:
The current logic would find it okay if any of the certs in the chain was correct in the certificate.

Would it be an idea to have, as part of the bundled set of CRLs, their original URL (so on the client-side)?
Then you can just build a map of URL->CRLs.
Then you could as part of this check, just look up the current CA's CRL URL, and look up the CRL in the map (returning a "No CRL provided" error if it didn't get sent along).
Then you don't have to "brute-force" finding a CA that's signed by the current CA: if it's not actually signed correctly, just return an error, and you know exactly which CRL to check for the current entry in the PKI path.

Reply:
The problem is that Intel provides three CRLs:

* CRL from CA cert
* Platform CRL via API
* CPU CRL via API

So those API CRLs have to be checked in the current manner I think.

Reply:
Added `(String, CertificateList)` structure, and the cert is checked by CRL identified by the CRL distribution point extension.